### ISSUE TITLE
Can not fetch cursor when executing multiple statements with AsyncConnection.execute()

### ISSUE THREAD

**Issue by mfrlin** [sqlalchemy/sqlalchemy#6174]
**Describe the bug**
I am trying to bulk insert multiple rows with one `AsyncConnection.execute()` call while also having `RETURNING` data. Rows do get inserted but I can not fetch cursor for data.

**Expected behavior**
Able to call `.fetchone()` or `.fetchall()` on cursor and get the data.

**To Reproduce**

First function inserts single row and works as expected. Second function throws exception.

```py
async def insert_document(db: AsyncConnection, storage_id: int, data: str):
    query = "INSERT INTO documents (storage_id, data) VALUES (:storage_id, :data) RETURNING id, updated_at, revision, data;"
    results = await db.execute(text(query), {"storage_id": storage_id, "data": data})
    return results.fetchone()
 
 
async def insert_documents(db: AsyncConnection, storage_id: int, data: List[str]):
    query = "INSERT INTO documents (storage_id, data) VALUES (:storage_id, :data) RETURNING id, updated_at, revision, data;"
    results = await db.execute(text(query), [{"storage_id": storage_id, "data": d} for d in data])
    # results.fetchone() or fetchall() causes sqlalchemy.exc.ResourceClosedError: This result object does not return rows. It has been closed automatically.
```

**Error**

```
my-project/models.py:61: in insert_documents
    results.fetchone()
../../.cache/pypoetry/virtualenvs/my-project-o7QgMDTb-py3.8/lib/python3.8/site-packages/sqlalchemy/engine/result.py:1010: in fetchone
    row = self._onerow_getter(self)
../../.cache/pypoetry/virtualenvs/my-project-o7QgMDTb-py3.8/lib/python3.8/site-packages/sqlalchemy/util/langhelpers.py:1160: in __get__
    obj.__dict__[self.__name__] = result = self.fget(obj)
../../.cache/pypoetry/virtualenvs/my-project-o7QgMDTb-py3.8/lib/python3.8/site-packages/sqlalchemy/engine/result.py:430: in _onerow_getter
    make_row = self._row_getter
../../.cache/pypoetry/virtualenvs/my-project-o7QgMDTb-py3.8/lib/python3.8/site-packages/sqlalchemy/util/langhelpers.py:1160: in __get__
    obj.__dict__[self.__name__] = result = self.fget(obj)
../../.cache/pypoetry/virtualenvs/my-project-o7QgMDTb-py3.8/lib/python3.8/site-packages/sqlalchemy/engine/result.py:320: in _row_getter
    keymap = metadata._keymap
../../.cache/pypoetry/virtualenvs/my-project-o7QgMDTb-py3.8/lib/python3.8/site-packages/sqlalchemy/engine/cursor.py:1197: in _keymap
    self._we_dont_return_rows()
../../.cache/pypoetry/virtualenvs/my-project-o7QgMDTb-py3.8/lib/python3.8/site-packages/sqlalchemy/engine/cursor.py:1178: in _we_dont_return_rows
    util.raise_(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

    def raise_(
        exception, with_traceback=None, replace_context=None, from_=False
    ):
        r"""implement "raise" with cause support.
    
        :param exception: exception to raise
        :param with_traceback: will call exception.with_traceback()
        :param replace_context: an as-yet-unsupported feature.  This is
         an exception object which we are "replacing", e.g., it's our
         "cause" but we don't want it printed.    Basically just what
         ``__suppress_context__`` does but we don't want to suppress
         the enclosing context, if any.  So for now we make it the
         cause.
        :param from\_: the cause.  this actually sets the cause and doesn't
         hope to hide it someday.
    
        """
        if with_traceback is not None:
            exception = exception.with_traceback(with_traceback)
    
        if from_ is not False:
            exception.__cause__ = from_
        elif replace_context is not None:
            # no good solution here, we would like to have the exception
            # have only the context of replace_context.__context__ so that the
            # intermediary exception does not change, but we can't figure
            # that out.
            exception.__cause__ = replace_context
    
        try:
>           raise exception
E           sqlalchemy.exc.ResourceClosedError: This result object does not return rows. It has been closed automatically.

../../.cache/pypoetry/virtualenvs/essence-datacenter-o7QgMDTb-py3.8/lib/python3.8/site-packages/sqlalchemy/util/compat.py:211: ResourceClosedError

```

**Versions.**
 - OS: Linux
 - Python: 3.8
 - SQLAlchemy: 1.4.4
 - Database: PostgreSQL
 - DBAPI: Asyncpg

**Comment by zzzeek** [comment]
hey there -

There's no support in any DBAPI right now for use of RETURNING at the same time as DBAPI.executemany().      What you can do instead is use INSERT with multiple VALUES like:

    stmt = "INSERT INTO table (x, y) VALUES (:x1, :y1), (:x2, :y2), (:x3, :y3), ..."
    result = conn.execute(stmt, {"x1": x1, "y1": y1, "x2": x2, "y2": y2, "x3": x3, "y3": y3,  ...}

that will also perform much better as all postgresql drivers are inefficient on executemany() calls.

there's more to this kind of thing within the PostgreSQL drivers where the psycopg2 driver can turn a set of parameters into the above VALUES form automatically, which SQLAlchemy seeks to generalize to the asyncpg and pg8000 DBAPIs also in an upcoming release, but this feature would require that you use the insert() object and not plain text in any case.   If you're using plain text, use the above form.

**Comment by mfrlin** [comment]
Thank you for a quick response, really helped me a lot. I couldn't find any info on this and am pretty much beginner in SQL world. Only semi related issue that I found is this one: https://github.com/sqlalchemy/sqlalchemy/issues/5433 but it didn't mention this exception.

I did as you suggested and it is roughly 4 times faster than inserting one by one.

Should I be vary of some pitfalls with this approach? In my case data is a string containing JSON documents. Is there a limit how "big" the statement can become, since those documents can be pretty large?
